pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java

```diff
@@ -0,0 +1,104 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.upsert;
+
+
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ComparisonColumns implements Comparable<ComparisonColumns> {
+  private Comparable[] _comparisonColumns;
+
+  public ComparisonColumns(Comparable[] comparisonColumns) {
+    _comparisonColumns = comparisonColumns;
+  }
+
+  public Comparable[] getComparisonColumns() {
+    return _comparisonColumns;
+  }
+
+  @Override
+  public int compareTo(ComparisonColumns other) {
+    // _comparisonColumns should only at most one non-null comparison value. If not, it is the user's responsibility.
+    // There is no attempt to guarantee behavior in the case where there are multiple non-null values
+    int comparisonResult;
+    int comparableIndex = getComparableIndex();
+
+    if (comparableIndex < 0) {
+      // All comparison values were null.  This record is only ok to keep if all prior values were also null
```

Comment:
Is this allowed? Should we just ignore this record (return -1)

Reply:
It's currently allowed in this change set in order to maintain parity with the case of a using a singular comparisonColumn.  Currently, if the comparison value is null it will take <default value> for the column.  If a subsequent record for the same PK also has a null comparison value, it will take <default value> for the column and end up with `compareTo` being `0` which results in the record being kept.

Allowing all values to be null when using multiple comparison columns would maintain this behaviour.  Thoughts?

Reply:
For singular comparison column case, comparison value should not be null, and null value can have undefined behavior because we just use the default value as the comparison value.
For multiple comparison columns case, there should be at least one comparison column not null. If one event does not have any non-null value, we can just drop it (unless there is no event associated with the key) since we don't know which value to compare.